### ISSUE TITLE
asynchronous method 'Quartz.Core.QuartzSchedulerThread.Run()', is waiting on a synchronous call to method 'System.Threading.Monitor.Wait(object, int)'

### ISSUE THREAD

**Issue by A9G-Data-Droid** [quartznet/quartznet#2354]
**Describe the bug**

I have a windows service I wrote in .NET8 that builds Quartz as a hosted service. It works great. It runs things on a schedule. When my service goes to stop it hangs indefinitely, never fully stopping. The final log I see is right before the quartz shutdown line. I have my services `ExecuteAsync` code wrapped in a try...catch with `await _scheduling.Shutdown().ConfigureAwait(false);` in the finally to ensure that this is the last thing it does before returning from `ExecuteAsync` where the host would handle the disposing of Quartz and everything else. 

I attached VS2022 debugger to the hung service, exported a dump, and analyzed it in VS2022. 

It says `AA0001    Thread pool thread or asnchronous task blocked on a synchronous call`

> 1 threads are performing asynchronous work in method 'Quartz.Core.QuartzSchedulerThread.Run()', but are waiting on a synchronous call to method 'System.Threading.Monitor.Wait(object, int)'. This may cause thread pool starvation and hangs.

Potential fix

> Do not synchronously wait on Monitors, Events, Tasks, or any other objects that may block your thread. See if you can update the method to be asynchronous.

**Version used**

Quartz (3.8.1)

**To Reproduce**

1. Follow [guide to make Windows Service in .NET Core](https://learn.microsoft.com/en-us/dotnet/core/extensions/windows-service)
2. Follow guide to [add Quartz to hosted service](https://www.quartz-scheduler.net/documentation/quartz-3.x/packages/hosted-services-integration.html#using)
3. Schedule a job
4. Start Quartz
5. Put `await _scheduling.Shutdown().ConfigureAwait(false);` at the end of your `ExecuteAsync` method inside your Windows service
6. Service is now hung and can't exit gracefully.

**Expected behavior**

`.Shutdown` should either complete or throw.

**Additional context**
The debugger help text is fairly clear that there is a problem with the async code pathways. Looking at [the code](https://github.com/quartznet/quartznet/blob/5f29a13c129c671d180014e05e0acc6bb318397c/src/Quartz/Core/QuartzSchedulerThread.cs#L471) for `Quartz.Core.QuartzSchedulerThread.Run` I do see the usage of `lock` and `Monitor.Wait(sigLock, timeUntilContinue);` which do not follow `async` programming as they are both synchronous. The new way to await a lock in async code is usually [SemaphoreSlim](https://learn.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim?view=net-8.0)

This issue appears to be the reason that [AsyncMonitor](https://github.com/StephenCleary/AsyncEx/blob/master/doc/AsyncMonitor.md)was written.

**Comment by lahma** [comment]
Would you like to submit a PR and propose some improvements?

**Comment by A9G-Data-Droid** [comment]
Sure! I am looking at it now.

**Comment by sherlock1982** [comment]
One easier solution will be to mark this task as `LongRunning` via `Task.Factory.StartNew`. In this case `ThreadPool` will not add extra tasks to it's queue.
That will still use the thread but will not trick `ThreadPool`. It will be almost as creating a dedicated thread

**Comment by jogibear9988** [comment]
@A9G-Data-Droid 
do you still work on this?

**Comment by A9G-Data-Droid** [comment]
> @A9G-Data-Droid 
> do you still work on this?

No. I am no longer using this library. I did make a PR at one point but there are too many changes required to fix it. The maintainer said they would implement some of the examples I gave in a comtrolled fashion. If you want to see what I was working on it's in this branch:

https://github.com/A9G-Data-Droid/quartznet/tree/Issue2354

If you need these things you could rebase and try to submit again with fewer changes, targeting only what you need to get the deadlock to stop.
